### PR TITLE
singleflight: add usage hint

### PR DESCRIPTION
**Pull request by andreimatei**

### CODE REVIEW

**Comment by andreimatei** [comment]
do you like this note?

File: pkg/util/syncutil/singleflight/singleflight.go [hunk 1]
```diff
@@ -70,10 +70,25 @@ func (g *Group) Do(
...
+// for.
+// 2. If it doesn't, initiate or join a flight to produce it.
+//
+// Step one is expected to be done while holding a lock. Modifying the
+// datastructure in the callback is expected to need to take the same lock. Once
+// a caller proceeds to step two, it likely wants to keep the lock until
+// DoChan() returned a channel, in order to ensure that a flight is only started
+// before any modifications to the datastructure occurred (relative to the state
+// observed in step one). Were the lock to be released before calling DoChan(),
+// a previous flight might modify the datastructure before our flight began.
```
Comment (nvanbenschoten) [inline pkg/util/syncutil/singleflight/singleflight.go]: The flight might also finish before we join it, meaning that we'll start a whole new one.

File: pkg/util/syncutil/singleflight/singleflight.go [hunk 2]
```diff
@@ -70,10 +70,25 @@ func (g *Group) Do(
...
-// DoChan is like Do but returns a channel that will receive the
-// results when they are ready. The method also returns a boolean
-// specifying whether the caller's fn function will be called or
-// not.
+// DoChan is like Do but returns a channel that will receive the results when
+// they are ready. The method also returns a boolean specifying whether the
+// caller's fn function will be called or not. This return value lets callers
+// identify a unique "leader" for a flight.
+//
+// NOTE: DoChan makes it possible to initiate or join a flight while holding a
```
Comment (nvanbenschoten) [inline pkg/util/syncutil/singleflight/singleflight.go]: Consider pointing to `RangeCache.LookupRangeDescriptor` as an example of this pattern.

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/syncutil/singleflight/singleflight.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19044#-KvdRyepJMCbb_1WN73g-r1-78:-Kw2FtK-mE4Ld2wESiu3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/67f57f8674ce1085c1b44531e6ea5cc1530f806e/pkg/util/syncutil/singleflight/singleflight.go#L78)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider pointing to `RangeCache.LookupRangeDescriptor` as an example of this pattern.
</blockquote></details>

Done.

---

*[pkg/util/syncutil/singleflight/singleflight.go, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19044#-KvdRyepJMCbb_1WN73g-r1-91:-Kw2Ftey05oPG_FFAgkj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/67f57f8674ce1085c1b44531e6ea5cc1530f806e/pkg/util/syncutil/singleflight/singleflight.go#L91)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The flight might also finish before we join it, meaning that we'll start a whole new one.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19044)*

**Review by nvb — APPROVED** [review]

**Pull request merged**